File: app/src/main/java/me/cafecode/android/newspaper/newses/NewsesPresenter.java
package me.cafecode.android.newspaper.newses;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import me.cafecode.android.newspaper.data.NewsRepository;
import me.cafecode.android.newspaper.data.models.News;

/**
 * Created by Natthawut Hemathulin on 5/31/2017 AD.
 * Email: natthawut1991@gmail.com
 */

public class NewsesPresenter implements NewsesContract.Presenter {

    private NewsesContract.View mView;

    private NewsRepository mRepository;

    @Inject
    public NewsesPresenter(NewsRepository repository, NewsesContract.View view) {
        mRepository = repository;
        mView = view;
    }

    @Override
    public void onStart() {
        mView.showProgressView(true);
    }

    @Override
    public void loadNewses() {
        mView.showProgressView(true);
        mRepository.loadNewses()
                .subscribe(new Observer<List<News>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<News> newses) {
                        mView.showNewses(newses);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {

                    }

                    @Override
                    public void onComplete() {
                        mView.showProgressView(false);
                    }
                });
    }

}

File: app/src/main/java/me/cafecode/android/newspaper/newses/NewsesPresenter.kt
package me.cafecode.android.newspaper.newses

import me.cafecode.android.newspaper.data.NewsRepository
import me.cafecode.android.newspaper.util.BaseSchedulerProvider
import javax.inject.Inject

class NewsesPresenter @Inject constructor(val repository: NewsRepository, val view:
NewsesContract.View, val schedulerProvider: BaseSchedulerProvider) : NewsesContract.Presenter {

    @Inject
    fun setupListeners() {
        view.setPresenter(this)
    }

    override fun onStart() {
        loadNewses()
    }

    override fun loadNewses() {
        view.showProgressView(true)

        repository.loadNewses()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        { newses ->
                            view.showNewses(newses)
                            view.showProgressView(false)
                        },
                        { error ->
                            error.printStackTrace()
                            view.showProgressView(false)
                            view.showErrorMessage(error.message!!)
                        }
                )
    }

}